tool/mustachio/codegen_aot_compiler.dart

```diff
@@ -0,0 +1,473 @@
+      writeln('}');
+    }
+  }
+
+  /// Compiles a value section containing [block] into a renderer's Dart source.
+  Future<void> _compileValueSection(
+      _VariableLookup variableLookup, List<MustachioNode> block,
+      {bool invert = false}) async {
+    var variableAccess = variableLookup.name;
+    if (invert) {
+      writeln('if ($variableAccess == null) {');
+      await _compile(block);
+      writeln('}');
+    } else {
+      var innerContextName = getNewContextName();
+      writeln('var $innerContextName = $variableAccess;');
+      writeln('if ($innerContextName != null) {');
+      var innerContext = _VariableLookup(variableLookup.type, innerContextName);
+      _contextStack.push(innerContext);
+      await _compile(block);
+      _contextStack.pop();
+      writeln('}');
+    }
+  }
+
+  /// Returns a valid [_VariableLookup] on a Mustache node, [node] by resolving
+  /// its key.
+  _VariableLookup _lookUpGetter(HasMultiNamedKey node) {
+    var key = node.key;
+
+    // '.' is an entirely special case.
+    if (key.length == 1 && key[0] == '.') {
+      return _VariableLookup(contextType, contextName);
+    }
+
+    var primaryName = key[0];
+
+    for (var context in _contextStack) {
+      var getter =
+          context.type.lookUpGetter2(primaryName, contextType.element.library);
+      if (getter == null) {
+        continue;
+      }
+
+      var type = getter.returnType;
+      var contextChain = '${context.name}.$primaryName';
+      var remainingNames = [...key.skip(1)];
+      for (var secondaryKey in remainingNames) {
+        getter = (type as InterfaceType)
+            .lookUpGetter2(secondaryKey, type.element.library);
+        if (getter == null) {
+          throw MustachioResolutionError(node.keySpan.message(
+              "Failed to resolve '$secondaryKey' on ${context.type} while "
+              'resolving $remainingNames as a property chain on any types in '
+              'the context chain: $contextChain, after first resolving '
+              "'$primaryName' to a property on $type"));
+        }
+        type = getter.returnType;
+        contextChain = '$contextChain.$secondaryKey';
+      }
+      return _VariableLookup(type, contextChain);
+    }
+
+    var contextTypes = [
+      for (var c in _contextStack) c.type,
+    ];
+    throw MustachioResolutionError(node.keySpan
+        .message("Failed to resolve '$key' as a property on any types in the "
+            'context chain: $contextTypes'));
+  }
+
+  /// Writes [content] to the generated render functions as text, properly
+  /// handling newlines, quotes, and other special characters.
+  void _writeText(String content) {
+    if (content.isEmpty) return;
+    if (content == '\n') {
+      // Blank lines happen a lot; just indicate them as such.
+      writeln('buffer.writeln();');
+    } else {
+      content = content.replaceAll("'", "\\'").replaceAll(r'$', r'\\$');
+      if (RegExp('^[ \\n]+\$').hasMatch(content)) {
```

Comment:
Suggest using a private static for the RegExp instead of constructing it each time, depending on how Dart handles it these constructions could be expensive.

Reply:
Great catch.